telluric/georaster.py

```diff
@@ -182,63 +254,69 @@ def _merge(one, other, merge_strategy=MergeStrategy.UNION, requested_bands=None)
-            new_bands = res_common.band_names
-            all_data = [res_common.image.data]
-
-        else:
-            res_common = one
-            new_bands = []
-            all_data = []
-
-        new_mask = res_common.image.mask[0]
-
-        # Add the rest of the bands
-        one_remaining_bands = [band for band in one.band_names if band not in set(common_bands)]
-        other_remaining_bands = [band for band in other.band_names if band not in set(common_bands)]
-
-        if one_remaining_bands:
-            all_data.insert(0, one.bands_data(one_remaining_bands).data)
-            # This is not necessary, as new_mask already includes
-            # at least all the values of one.image.mask because it comes
-            # either from one or from the intersection of one and other
-            # new_mask |= one.image.mask[0]
-            new_bands = one_remaining_bands + new_bands
-
-        if other_remaining_bands:
-            all_data.append(other.bands_data(other_remaining_bands).data)
-            # Apply "or" to the mask in the same way rasterio does, see
-            # https://mapbox.github.io/rasterio/topics/masks.html#dataset-masks
-            new_mask |= other.image.mask[0]
-            new_bands = new_bands + other_remaining_bands
-        new_image = np.ma.MaskedArray(
-            np.concatenate(all_data),
-            mask=[new_mask] * len(new_bands)
+        assert set(one.band_names).intersection(set(other.band_names)) == set()
+
+        # We raise an error in the bands are the same. See above.
+        if one.band_names == other.band_names:
+            raise ValueError("rasters have the same bands, use another merge strategy")
+
+        # Apply "or" to the mask in the same way rasterio does, see
+        # https://mapbox.github.io/rasterio/topics/masks.html#dataset-masks
+        # In other words, mask the values that are already masked in either
+        # of the two rasters, since one mask per band is not supported
+        new_mask = one.image.mask[0] | other.image.mask[0]
+
+        # Concatenate the data along the band axis and apply the mask
+        new_image = np.ma.masked_array(
+            np.concatenate([
+                one.image.data,
+                other.image.data
+            ]),
+            mask=[new_mask] * (one.image.shape[0] + other.image.shape[0])
         )
+        new_bands = one.band_names + other.band_names
+
         # We don't copy image and mask here, due to performance issues,
         # this output should not use without eventually being copied
         # In this context we are copying the object in the end of merge_all merge_first and merge
         return _Raster(image=new_image, band_names=new_bands)
+
     else:
         raise ValueError("Use one the strategies available in MergeStrategy instead.")
 
 
+def _merge_two(one, other, merge_strategy=MergeStrategy.UNION, silent=False):
```

Comment:
no documentation what `_merge_two` does why we need this method

So I can't review this code, and to see that you get what you expected

Reply:
The docstring is in `merge` and `merge_to_first`. I also don't know why we are keeping those cases, just made the code maintain the current behavior.